File: scripts/hooks/lib/record-component.js
// Shared writer for component_outcomes: the PostToolUse recorder and every
// guard's deny path go through here. Never throws, because recording must never
// change what the caller was going to do.

const fs = require('fs');

const DEFAULT_DB = 'C:/Users/skf_s/.claude/dev-framework/episodes.db';

function record({ kind, name, sessionId, cwd, blocked = 0, notes = null }) {
  try {
    if (!kind || !name) return;

    const dbPath = process.env.DEVRL_DB || DEFAULT_DB;
    // Opening a missing file would create an empty db in whatever repo we ran in.
    if (!fs.existsSync(dbPath)) return;

    process.removeAllListeners('warning');
    process.on('warning', () => {});

    const { DatabaseSync } = require('node:sqlite');
    const db = new DatabaseSync(dbPath);
    try {
      db.exec('PRAGMA busy_timeout = 3000');
      db.prepare(
        'INSERT INTO component_outcomes ' +
          '(kind, name, session_id, invoked_at, blocked, cwd, notes) ' +
          'VALUES (?, ?, ?, ?, ?, ?, ?)'
      ).run(
        String(kind),
        String(name),
        sessionId || null,
        new Date().toISOString(),
        blocked ? 1 : 0,
        cwd || null,
        notes ? String(notes).slice(0, 500) : null
      );
    } finally {
      db.close();
    }
  } catch {
    // Swallowed on purpose: every failure here must leave the caller untouched.
  }
}

module.exports = { record };

File: scripts/hooks/lib/record_component.py
"""Shared writer for component_outcomes, the Python twin of record-component.js.

Every .py hook's deny path goes through here. Never raises: recording must not
change what the caller was going to do.
"""

from __future__ import annotations

import os
import sqlite3
from datetime import datetime, timezone
from pathlib import Path

DEFAULT_DB = Path(__file__).resolve().parents[3] / "dev-framework" / "episodes.db"


def record(
    *,
    kind: str,
    name: str,
    session_id: str | None = None,
    cwd: str | None = None,
    blocked: bool = False,
    notes: str | None = None,
) -> None:
    try:
        if not kind or not name:
            return
        db = Path(os.environ.get("DEVRL_DB") or DEFAULT_DB)
        if not db.is_file():  # opening a missing file would create an empty db in any repo
            return
        conn = sqlite3.connect(db, timeout=3)
        try:
            conn.execute(
                "INSERT INTO component_outcomes "
                "(kind, name, session_id, invoked_at, blocked, cwd, notes) "
                "VALUES (?, ?, ?, ?, ?, ?, ?)",
                (
                    str(kind),
                    str(name),
                    session_id,
                    datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"),
                    1 if blocked else 0,
                    cwd,
                    str(notes)[:500] if notes else None,
                ),
            )
            conn.commit()
        finally:
            conn.close()
    except Exception:
        pass
